Html preview

ferdinandyb opened the pull request:
<!-- Provide a descriptive summary of the changes in the title above -->

#### ISSUE TYPE
<!-- Pick relevant types and delete the rest -->
- Bug fix
- Improvement/feature implementation
- Breaking changes

#### RUNTIME ENVIRONMENT
<!-- Details of your runtime environment -->
<!-- Retrieve Python/ranger version and locale with `ranger -\-version` -->
- Operating system and version: 
- Terminal emulator and version: 
- Python version: 
- Ranger version/commit: 
- Locale: 

#### CHECKLIST
<!-- All [REQUIRED] requisites need to be fulfilled -->
<!-- Replace [ ] with [X] when fulfilled -->
- [x] The `CONTRIBUTING` document has been read **[REQUIRED]**
- [ ] All changes follow the code style **[REQUIRED]**
- [ ] All new and existing tests pass **[REQUIRED]**
- [ ] Changes require config files to be updated
    - [ ] Config files have been updated
- [ ] Changes require documentation to be updated
    - [ ] Documentation has been updated
- [ ] Changes require tests to be updated
    - [ ] Tests have been updated

#### DESCRIPTION
Adds an optional image preview for html based on `wkhtmltoimage` and documents this dependency in the README.


#### MOTIVATION AND CONTEXT
<!-- Why are these changes required? -->
<!-- What problems do these changes solve? -->
Looks better.
<!-- Link to relevant issues -->


#### TESTING
<!-- What tests have been run? -->
<!-- How does the changes affect other areas of the codebase? -->
Shouldn't affect anything.

#### IMAGES / VIDEOS<!-- Only if relevant -->
<!-- Link or embed images and videos of screenshots, sketches etc. -->
![image](https://user-images.githubusercontent.com/6343487/146690460-5443a27c-5a06-4317-baf7-bca46fa14f21.png)

ferdinandyb committed:
adding image preview for html via wkhtmltoimage

ferdinandyb committed:
placing everything in comments so default preview is still text

ferdinandyb commented:
I think I messed up something with creating the PR (and I thought I'd get an option from github to squash my commits?), but it this PR adds an optional image preview of html files, a bit akin to how pdfs can be now previewed, based on the suggest on [stackoverflow](https://unix.stackexchange.com/a/683004/82390).

toonn reviewed (dismissed):
LGTM, will try to get around to testing it soon. Feel free to ping if I take more than let's say 2 weeks.

ferdinandyb commented:
great, thanks!

ferdinandyb commented:
A friendly ping :)

toonn committed:
scope: Change html image preview to use JPEG

We still hardcode the `.jpg` extension for cached previews so it's nicer
not to pretend a PNG is a JPEG.

toonn committed:
scope: HTML image preview fix crop height

toonn committed:
scope: Split long line

toonn commented:
I really like this and I can't really see anything wrong with the implementation but I cannot get it working. The preview is generated just fine in the cache but ranger fails to display it. Can anyone else confirm this working?

(I made some minor changes. The height specification wasn't quite correct, though the default wide size isn't ideal for webpages.)

ferdinandyb commented:
It works for me. What I saw, on presumably larger files, is that it felt like I needed to go off the item and back to get the preview, but maybe I was just impatient.

ferdinandyb commented:
Could there be a difference between the html-s we are trying to convert? Could you send me the file where this fails?

markus-bauer commented:
I was actually using the python documentation which can be downloaded here:
https://docs.python.org/3/download.html
zip:
https://docs.python.org/3/archives/python-3.10.4-docs-html.zip


I tried it on a simpler html file, that doesn't use any other resources (like css, images, js). This worked.

I then just added an image tag to that simple html, and it failed with the following error (but it renders in firefox):

```
Loading page (1/2)
Warning: Blocked access to file /home/.../test.jpg
Error: Failed to load about:blank, with network status code 301 and http status code 0 - Protocol "about" is unknown
Rendering (2/2)                                                    
Done                                                               
Exit with code 1 due to network error: ProtocolUnknownError
```
The python doc failed similarly with Blocked access to css/js.

So it's probably related to loading of additional local resources.

But I don't have any more time for this today. If there are other questions, I will respond tomorrow later this week.
